crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+};
+use restate_types::{
+    identifiers::PartitionId,
+    net::ingress::{IngestRecord, IngestRequest, IngestResponse},
+    retries::{RetryIter, RetryPolicy},
+};
+
+use crate::chunks_timeout::ChunksTimeout;
+
+/// Error returned when attempting to use a session that has already been closed.
+#[derive(Clone, Copy, Debug, thiserror::Error)]
+#[error("Partition session is closed")]
+pub struct SessionClosed;
+
+/// Commitment failures that can be observed when waiting on [`RecordCommit`].
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub enum CommitError {
+    #[error("commit cancelled")]
+    Cancelled,
+}
+
+/// Future that is resolved to the commit result
+/// A [`CommitError::Cancelled`] might be returned
+/// if [`crate::Ingress`] is closed while record is in
+/// flight. This does not guarantee that the record
+/// was ont processed or committed.
+pub struct RecordCommit {
+    rx: oneshot::Receiver<Result<(), CommitError>>,
+}
+
+impl Future for RecordCommit {
+    type Output = Result<(), CommitError>;
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        match ready!(self.rx.poll_unpin(cx)) {
+            Ok(result) => std::task::Poll::Ready(result),
+            Err(_) => std::task::Poll::Ready(Err(CommitError::Cancelled)),
+        }
+    }
+}
+
+impl RecordCommit {
+    fn new() -> (Self, RecordCommitResolver) {
+        let (tx, rx) = oneshot::channel();
+        (Self { rx }, RecordCommitResolver { tx })
+    }
+}
+
+struct RecordCommitResolver {
+    tx: oneshot::Sender<Result<(), CommitError>>,
+}
+
+impl RecordCommitResolver {
+    /// Resolve the [`RecordCommit`] to committed.
+    pub fn committed(self) {
+        let _ = self.tx.send(Ok(()));
+    }
+
+    /// explicitly cancel the RecordCommit
+    /// If resolver is dropped, the RecordCommit
+    /// will resolve to [`CommitError::Cancelled`]
+    #[allow(dead_code)]
+    pub fn cancelled(self) {
+        let _ = self.tx.send(Err(CommitError::Cancelled));
+    }
+}
+
+struct IngressBatch {
+    records: Arc<[IngestRecord]>,
+    trackers: Vec<RecordTracker>,
+    reply_rx: Option<ReplyRx<IngestResponse>>,
+}
+
+impl IngressBatch {
+    /// Marks every tracked record in the batch as committed.
+    pub fn committed(self) {
+        for tracker in self.trackers {
+            tracker.resolver.committed();
+        }
+    }
+}
+
+/// Tunable parameters for batching and networking behaviour of partition sessions.
+#[derive(Debug, Clone)]
+pub struct SessionOptions {
+    /// Maximum batch size
+    pub batch_size: usize,
+    /// Maximum batch timeout before it's sent
+    pub batch_timeout: Duration,
+    /// connection retry policy
+    pub connect_retry_policy: RetryPolicy,
+}
+
+impl Default for SessionOptions {
+    fn default() -> Self {
+        Self {
+            batch_size: 250,
+            batch_timeout: Duration::from_millis(100),
```

Comment:
How did you come up with these default values? What happens if I have 250 very large items in my batch?

Reply:
Totally arbitrary, I will see how to change this to be in bytes instead.